### PR TITLE
[Placement Group] Support RayPlacementGroupError #10508

### LINKED ISSUE #10508: [Placement Group] Support RayPlacementGroupError
### Describe your feature request
When tasks are failed due to placement group removal, we should raise `RayPlacementGroupError` instead of worker dead error or actor dead error.

### PR DESCRIPTION
**Pull request by oliverhu**
## Why are these changes needed?
Fix https://github.com/ray-project/ray/issues/10508

Also:
- Refactored `ProcessDisconnectClientMessage` to an msg processing function and an util function. A lot of message passing is from the same file.
- Removed `intentional_disconnect_client`, replaced with a `ClientDisconnectType` to provide more information on disconnect reasoning.
- Renamed `DestroyWorker` to `DisconnectAndKillWorker` to bring more clarity.

## Related issue number
https://github.com/ray-project/ray/issues/10508

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by oliverhu** [comment]
```>>> import ray
>>> ray.init()
2020-12-12 00:56:41,410	INFO services.py:1171 -- View the Ray dashboard at http://127.0.0.1:8265
[2020-12-12 00:56:42,028 I 48095 12603168] logging.cc:272: Set ray log level from environment variable RAY_BACKEND_LOG_LEVEL to -1
{'node_ip_address': '192.168.0.10', 'raylet_ip_address': '192.168.0.10', 'redis_address': '192.168.0.10:36109', 'object_store_address': '/tmp/ray/session_2020-12-12_00-56-40_791646_48095/sockets/plasma_store', 'raylet_socket_name': '/tmp/ray/session_2020-12-12_00-56-40_791646_48095/sockets/raylet', 'webui_url': '127.0.0.1:8265', 'session_dir': '/tmp/ray/session_2020-12-12_00-56-40_791646_48095', 'metrics_export_port': 60347, 'node_id': '58021499d0338a82e82eadad08ef963c4f878bcb'}
>>> from ray.util.placement_group import (
...      placement_group,
...      placement_group_table,
...      remove_placement_group
... )
>>> pg = placement_group([{"CPU": 2}])
>>> ray.get(pg.ready())
<ray.util.placement_group.PlacementGroup object at 0x7fb7801eec70>
>>>
>>> @ray.remote(num_cpus=2)
... def f():
...    import time
...    time.sleep(20)
...    return True
...
>>> f.options(placement_group=pg).remote()
ObjectRef(cb230a572350ff44ffffffff0100000001000000)
>>> remove_placement_group(pg)
>>> 2020-12-12 00:56:46,418	WARNING worker.py:1032 -- A worker was killed while executing task cb230a572350ff44ffffffff01000000 due to placement group error..
(pid=48139) 2020-12-12 00:56:46,416	ERROR worker.py:383 -- SystemExit was raised from the worker
(pid=48139) Traceback (most recent call last):
(pid=48139)   File "python/ray/_raylet.pyx", line 551, in ray._raylet.task_execution_handler
(pid=48139)   File "python/ray/_raylet.pyx", line 418, in ray._raylet.execute_task
(pid=48139)   File "python/ray/_raylet.pyx", line 457, in ray._raylet.execute_task
(pid=48139)   File "python/ray/_raylet.pyx", line 460, in ray._raylet.execute_task
(pid=48139)   File "python/ray/_raylet.pyx", line 464, in ray._raylet.execute_task
(pid=48139)   File "<stdin>", line 4, in f
(pid=48139)   File "/Users/khu/ray/ray/python/ray/worker.py", line 380, in sigterm_handler
(pid=48139)     sys.exit(1)
(pid=48139) SystemExit: 1
(pid=raylet) [2020-12-12 00:56:46,517 E 48107 12603430] process.cc:498: Failed to kill process 48139 with error system:3: No such process
>>> 2020-12-12 00:56:51,999	WARNING worker.py:1032 -- The actor or task with ID cb230a572350ff44ffffffff01000000 cannot be scheduled right now. It requires {CPU_group_8f90f3e8dbdd33c5819fb5f8244503e9: 2.000000} for placement, however the cluster currently cannot provide the requested resources. The required resources may be added as autoscaling takes place or placement groups are scheduled. Otherwise, consider reducing the resource requirements of the task.

>>>
```

File: src/ray/raylet/node_manager.cc [hunk 1]
```diff
@@ -1377,27 +1374,35 @@ void NodeManager::ProcessDisconnectClientMessage(
...
+    std::ostringstream error_message;
+    if (!intentional_disconnect) {
+      type_str = "worker_died";
+      error_message << "A worker died or was killed while executing task " << task_id
+                    << ".";
+    } else {
+      error_type = ErrorType::PLACEMENT_GROUP_ERROR;
+      type_str = "placement_group_error";
+      error_message << "A worker was killed while executing task " << task_id
+                    << " due to placement group error"
```
Comment (rkooo567) [inline src/ray/raylet/node_manager.cc:1389]: Maybe we can call it due to placement group removal?

File: src/ray/raylet/node_manager.cc [hunk 2]
```diff
@@ -1104,13 +1106,7 @@ void NodeManager::ProcessClientMessage(const std::shared_ptr<ClientConnection> &
...
   } break;
   case protocol::MessageType::DisconnectClient: {
-    ProcessDisconnectClientMessage(client);
-    // We don't need to receive future messages from this client,
-    // because it's already disconnected.
-    return;
-  } break;
-  case protocol::MessageType::IntentionalDisconnectClient: {
-    ProcessDisconnectClientMessage(client, /* intentional_disconnect = */ true);
+    ProcessDisconnectClientMessage(client, message_data);
```
Comment (rkooo567) [inline src/ray/raylet/node_manager.cc:1109]: Why are we deleting the intentional disconnect check here?

**Comment by rkooo567** [comment]
The PR looks great! Have a few questions (and I also want to see if our CI passes properly!)

File: src/ray/raylet/node_manager.cc (hunk 2 above)
Comment (oliverhu) [reply src/ray/raylet/node_manager.cc:1109]: Created a disconnect type payload in `message_data`

**Comment by oliverhu** [comment]
```
INFO: Build completed successfully, 19 total actions
//:actor_manager_test                                           (cached) PASSED in 18.6s
//:asio_test                                                    (cached) PASSED in 2.0s
//:client_connection_test                                       (cached) PASSED in 0.9s
//:core_worker_test                                             (cached) PASSED in 30.7s
//:create_request_queue_test                                    (cached) PASSED in 0.2s
//:direct_actor_transport_test                                  (cached) PASSED in 30.9s
//:direct_task_transport_test                                   (cached) PASSED in 0.3s
//:event_test                                                   (cached) PASSED in 3.5s
//:filesystem_test                                              (cached) PASSED in 0.2s
//:gcs_actor_manager_test                                       (cached) PASSED in 0.2s
//:gcs_actor_scheduler_test                                     (cached) PASSED in 0.3s
//:gcs_node_manager_test                                        (cached) PASSED in 0.3s
//:gcs_object_manager_test                                      (cached) PASSED in 0.2s
//:gcs_placement_group_manager_test                             (cached) PASSED in 0.9s
//:gcs_placement_group_scheduler_test                           (cached) PASSED in 0.6s
//:gcs_pub_sub_test                                             (cached) PASSED in 5.1s
//:gcs_resource_manager_test                                    (cached) PASSED in 0.2s
//:gcs_server_rpc_test                                          (cached) PASSED in 3.1s
//:gcs_server_test                                              (cached) PASSED in 69.0s
//:global_state_accessor_test                                   (cached) PASSED in 6.8s
//:id_test                                                      (cached) PASSED in 0.2s
//:in_memory_gcs_table_storage_test                             (cached) PASSED in 0.5s
//:in_memory_store_client_test                                  (cached) PASSED in 0.6s
//:logging_test                                                 (cached) PASSED in 0.2s
//:metric_exporter_client_test                                  (cached) PASSED in 0.5s
//:object_recovery_manager_test                                 (cached) PASSED in 0.3s
//:redis_actor_info_accessor_test                               (cached) PASSED in 0.6s
//:redis_gcs_client_test                                        (cached) PASSED in 6.6s
//:redis_gcs_table_storage_test                                 (cached) PASSED in 0.8s
//:redis_job_info_accessor_test                                 (cached) PASSED in 0.6s
//:redis_node_info_accessor_test                                (cached) PASSED in 1.2s
//:redis_object_info_accessor_test                              (cached) PASSED in 0.6s
//:redis_store_client_test                                      (cached) PASSED in 2.0s
//:reference_count_test                                         (cached) PASSED in 0.3s
//:sample_test                                                  (cached) PASSED in 0.5s
//:scheduling_queue_test                                        (cached) PASSED in 30.3s
//:scheduling_resources_test                                    (cached) PASSED in 0.2s
//:sequencer_test                                               (cached) PASSED in 0.7s
//:stats_test                                                   (cached) PASSED in 3.4s
//:task_manager_test                                            (cached) PASSED in 1.0s
//:util_test                                                    (cached) PASSED in 1.1s
//:cluster_resource_scheduler_test                                       PASSED in 0.7s
//:cluster_task_manager_test                                             PASSED in 0.7s
//:local_object_manager_test                                             PASSED in 0.9s
//:push_manager_test                                                     PASSED in 0.8s
//:reconstruction_policy_test                                            PASSED in 2.1s
//:signal_test                                                           PASSED in 1.6s
//:task_dependency_manager_test                                          PASSED in 0.9s
//:worker_pool_test                                                      PASSED in 0.9s
```

File: python/ray/exceptions.py [hunk 3]
```diff
@@ -138,6 +138,13 @@ def __str__(self):
         return "\n".join(out)
 
 
+class RayPlacementGroupError(RayError):
+    """Indicates that the task failed due to placement errors."""
```
Comment (ffbin) [inline python/ray/exceptions.py:142]:
```suggestion
    """Indicates that the task failed due to placement group error."""
```

File: python/ray/exceptions.py [hunk 4]
```diff
@@ -138,6 +138,13 @@ def __str__(self):
...
+
+    def __str__(self):
+        return "The task failed due to placement group errors."
```
Comment (ffbin) [inline python/ray/exceptions.py:145]:
```suggestion
        return "The task failed due to placement group error."
```

File: python/ray/ray_constants.py [hunk 5]
```diff
@@ -130,6 +130,7 @@ def to_memory_units(memory_bytes, round_up):
 DASHBOARD_AGENT_DIED_ERROR = "dashboard_agent_died"
 DASHBOARD_DIED_ERROR = "dashboard_died"
 RAYLET_CONNECTION_ERROR = "raylet_connection_error"
+PLACEMENT_GROUP_PUSH_ERROR = "placement_group_error"
```
Comment (ffbin) [inline python/ray/ray_constants.py:133]:
```suggestion
PLACEMENT_GROUP_PUSH_ERROR = "placement_group"
```

File: src/ray/gcs/gcs_server/gcs_actor_manager.h [hunk 6]
```diff
@@ -250,7 +250,8 @@ class GcsActorManager : public rpc::ActorInfoHandler {
   /// \param intentional_exit Whether the death was intentional. If yes and the
```
Comment (ffbin) [inline src/ray/gcs/gcs_server/gcs_actor_manager.h:250]: Modify comments synchronously?

File: src/ray/protobuf/common.proto [hunk 7]
```diff
@@ -133,6 +133,8 @@ enum ErrorType {
   TASK_CANCELLED = 5;
   // Inidicates that creating the GCS service failed to create the actor.
   ACTOR_CREATION_FAILED = 6;
+  // Indicates failure due to placementgroup
```
Comment (ffbin) [inline src/ray/protobuf/common.proto:136]:
```suggestion
  // Indicates failure due to placement group.
```

File: src/ray/protobuf/common.proto [hunk 8]
```diff
@@ -437,3 +439,15 @@ message MetricPoint {
   // [Optional] Unit of the metric.
   string units = 6;
 }
+
+// Type of a worker.
```
Comment (ffbin) [inline src/ray/protobuf/common.proto:443]:
```suggestion
// Type of a worker exit.
```

File: src/ray/protobuf/common.proto [hunk 9]
```diff
@@ -437,3 +439,15 @@ message MetricPoint {
...
+enum ClientDisconnectType {
+  // Intended worker exit.
+  FINISHED = 0;
+  // Unintended worker exit.
+  UNEXPECTED_EXIT = 1;
+  // Worker exit due to resource bundle release.
+  RELEASE_UNUSED_RESOURCE = 2;
```
Comment (ffbin) [inline src/ray/protobuf/common.proto:450]: UNUSED_RESOURCE_RELEASE?

File: src/ray/protobuf/gcs.proto [hunk 10]
```diff
@@ -400,7 +400,7 @@ message WorkerTableData {
   // The UNIX timestamp at which this worker's state was updated.
   int64 timestamp = 3;
   // Whether it's an intentional disconnect, only applies then `is_alive` is false.
```
Comment (ffbin) [inline src/ray/protobuf/gcs.proto:402]: Modify comments synchronously.

**Comment by rkooo567** [comment]
It looks like `test_failure` is broken!

File: src/ray/raylet/node_manager.cc [hunk 11]
```diff
@@ -331,11 +331,14 @@ void NodeManager::KillWorker(std::shared_ptr<WorkerInterface> worker) {
   });
 }
 
-void NodeManager::DestroyWorker(std::shared_ptr<WorkerInterface> worker) {
+void NodeManager::DisconnectAndKillWorker(std::shared_ptr<WorkerInterface> worker,
+                                          rpc::ClientDisconnectType disconnect_type) {
+  // Used to destroy workers when its the bundle resource is released (unused or
```
Comment (ffbin) [inline src/ray/raylet/node_manager.cc:336]:
```suggestion
  // Used to destroy worker when its bundle resource is released (unused or
```

File: src/ray/raylet/node_manager.cc [hunk 12]
```diff
@@ -1368,36 +1363,45 @@ void NodeManager::ProcessDisconnectClientMessage(
...
+      ErrorType error_type = ErrorType::WORKER_DIED;
+      // TODO(rkn): Define this constant somewhere else.
+      std::string type_str;
+      std::ostringstream error_message;
+      if (disconnect_type == rpc::ClientDisconnectType::UNEXPECTED_EXIT) {
+        type_str = "worker_died";
+        error_message << "A worker died or was killed while executing task " << task_id
+                      << ".";
+      }
+      if (disconnect_type == rpc::ClientDisconnectType::PLACEGROUP_REMOVED) {
```
Comment (ffbin) [inline src/ray/raylet/node_manager.cc:1386]: } else if (disconnect_type == rpc::ClientDisconnectType::PLACEGROUP_REMOVED) {

File: src/ray/raylet/node_manager.cc [hunk 13]
```diff
@@ -1368,36 +1363,45 @@ void NodeManager::ProcessDisconnectClientMessage(
...
-        std::ostringstream error_message;
-        error_message << "A worker died or was killed while executing task " << task_id
-                      << ".";
-        auto error_data_ptr = gcs::CreateErrorTableData(type, error_message.str(),
-                                                        current_time_ms(), job_id);
-        RAY_CHECK_OK(gcs_client_->Errors().AsyncReportJobError(error_data_ptr, nullptr));
-      }
+      // Push the error to driver.
+      const JobID &job_id = worker->GetAssignedJobId();
+      auto error_data_ptr = gcs::CreateErrorTableData(type_str, error_message.str(),
```
Comment (ffbin) [inline src/ray/raylet/node_manager.cc:1402]:
```suggestion
      const auto &error_data_ptr = gcs::CreateErrorTableData(type_str, error_message.str(),
```

File: src/ray/raylet/node_manager.h [hunk 14]
```diff
@@ -429,12 +429,16 @@ class NodeManager : public rpc::NodeManagerServiceHandler {
...
   /// We will disconnect the worker connection first and then kill the worker.
   ///
-  /// \param worker The worker to destroy.
+  /// \param worker The worker to stop.
   /// \return Void.
-  void DestroyWorker(std::shared_ptr<WorkerInterface> worker);
+  void DisconnectAndKillWorker(std::shared_ptr<WorkerInterface> worker,
+                               rpc::ClientDisconnectType disconnect_type);
+
+  void DisconnectClient(const std::shared_ptr<ClientConnection> &client,
```
Comment (ffbin) [inline src/ray/raylet/node_manager.h:440]: Pls add function description.

File: python/ray/ray_constants.py (hunk 5 above)
Comment (oliverhu) [reply python/ray/ray_constants.py:133]: should be place_group_error?

File: src/ray/raylet/node_manager.cc (hunk 13 above)
Comment (oliverhu) [reply src/ray/raylet/node_manager.cc:1402]: Is this the recommended way to declare erro_data_ptr? it is `auto error_data_ptr ` in the code base

**Comment by oliverhu** [comment]
@rkooo567 all tests pass for me locally tho...ran a few times

File: python/ray/ray_constants.py (hunk 5 above)
Comment (ffbin) [reply python/ray/ray_constants.py:133]: I found the others don't have `error`, for example:`REGISTER_ACTOR_PUSH_ERROR`. @rkooo567 What's your suggestion?

File: src/ray/raylet/node_manager.cc (hunk 13 above)
Comment (ffbin) [reply src/ray/raylet/node_manager.cc:1401]: `const auto &error_data_ptr` can reduce one copy. Actually, I don't quite understand why was `auto error_data_ptr` before.

Comment (rkooo567) [reply src/ray/raylet/node_manager.cc:1401]: The return value is actually made inside CreateErrorTableData though? I think we should just do const auto error_data_ptr? (also the return value is anyway a pointer)

File: python/ray/ray_constants.py (hunk 5 above)
Comment (rkooo567) [reply python/ray/ray_constants.py:133]: Yeah naming conventions here are not that consistent in general... I prefer to follow the majority (no error).

**Review by rkooo567 — APPROVED** [review]
LGTM if tests pass

File: src/ray/raylet/node_manager.cc (hunk 13 above)
Comment (ffbin) [reply src/ray/raylet/node_manager.cc:1401]: SGTM

**Review by ffbin — APPROVED** [review]
LGTM

File: src/ray/raylet/node_manager.cc (hunk 13 above)
Comment (oliverhu) [reply src/ray/raylet/node_manager.cc:1401]: updated to const auto everywhere!

File: python/ray/ray_constants.py (hunk 5 above)
Comment (oliverhu) [reply python/ray/ray_constants.py]: updated, also updated `raylet_connection_error`

**Comment by oliverhu** [comment]
@rkooo567 is right, ./python/ray/tests/test_failure.py is failing, not sure why not captured in `bazel test //:all`

**Comment by oliverhu** [comment]
Found the issue to we changed the fault err message passing from UNEXPECTED to EXPECTED, implicit is 0. Flipped the order.

**Comment by oliverhu** [comment]
_In the meantime, please don't merge_

**Comment by rkooo567** [comment]
As discussed, let's isolate the refactoring part and work on it in the next PR.

**Comment by oliverhu** [comment]
closing this PR to create a new one w/o the refactoring changes.

**Pull request closed without merging**